### PR TITLE
Add support to display current charge, and indicate whether or not device is charging

### PR DESCRIPTION
**Pull request by danielbernard**
- Added menuExit function to handle deinit for psm
- Not certain how you want that handled for PC, so psm calls are isolated to switch configuration
- Updated Makefile.pc to incorporate new icon binaries

light_theme_charging
light_theme_not_charging
dark_theme_charging
dark_theme_not_charging

### CODE REVIEW

**Comment by danielbernard** [comment]
Ah shoot, I didn't check the theme menu, so the charge text overlaps the "Theme Menu" text.
Would you prefer the text to be smaller, or maybe inline with the time?
2018100710265100-db1426d1dfd034027cecde9c2dd914b8

**Comment by danielbernard** [comment]
Example of smaller text to fit between the "Theme Menu" and time:
2018100710265100-db1426d1dfd034027cecde9c2dd914b8

**Comment by danielbernard** [comment]
Example of smaller text to fit between the "Theme Menu" and time:
2018100711032800-db1426d1dfd034027cecde9c2dd914b8

**Comment by danielbernard** [comment]
I definitely misunderstood the theme mechanisms that were employed, apologies for my bad read. 
I'll rework this to use the current theme in some way.

Would it be preferred to make the icon be an individually configurable color, or make it match the theme's textColor?

Is there a location I can see what symbols are available in the sharedfont? I'm having trouble finding a copy.
If there isn't a reasonably desired symbol that matches, I'll try to find a way to render the shape nicely.

**Comment by yellows8** [comment]
There's no symbol in sharedfont for this. "match the theme's textColor" Doesn't matter. That draw-code is missing a check for the {is-initialized} flag.

**Comment by danielbernard** [comment]
Would it be helpful to have like a "drawIcon" function that operated similarly to drawImage but forced the color to a specified value?

File: common/menu.c [hunk 1]
```diff
@@ -406,6 +440,33 @@ void drawTime() {
 
 }
 
+void drawCharge() {
+
+    char chargeString[5];
+
+    uint32_t batteryCharge;
+    ChargerType chargeType = ChargerType_None;
```
Comment (yellows8) [inline common/menu.c]: Breaks pc-build.

Comment (danielbernard) [reply common/menu.c]: How does changing this to the following looks?
```
    #ifdef __SWITCH__
    ChargerType chargeType = ChargerType_None;
    if (psmInitialized)
    {
        psmGetBatteryChargePercentage(&batteryCharge);
        psmGetChargerType(&chargeType);
    }
    #else
    batteryCharge = 100;
    #endif

    ...

    #ifdef __SWITCH__
    if (chargeType > (ChargerType)ChargerType_None)
    #else
    if (false) // No action for non-switch devices
    #endif
        drawImage(tmpX - 20, 0 + 47 + 10 + 5, 9, 15, charging_icon_small, IMAGE_MODE_RGBA32);
```

**Comment by yellows8** [comment]
"drawIcon" sure

**Comment by danielbernard** [comment]
Refactored, let me know if that approach is acceptable.

**Comment by yellows8** [comment]
Screenshot with current changes?

**Comment by danielbernard** [comment]
Using the following as a theme:
```
themeInfo={
    name="feature/charge-display Theme";
    author="ReckedCat";
    version="1.0.0";
};

lightTheme={
    textColor=(200,0,0,255);
    frontWaveColor=(255,102,204,255);
    middleWaveColor=(255,51,153,255);
    backWaveColor=(204,0,153,255);
    backgroundColor=(233,236,241,255);
    highlightColor=(255,77,166,255);
    separatorColor=(128,0,128,255);
    enableWaveBlending=0;
};

darkTheme={
    textColor=(0,255,255,255);
    frontWaveColor=(255,51,153,255);
    middleWaveColor=(255,102,204,255);
    backWaveColor=(204,0,153,255);
    backgroundColor=(0,0,0,255);
    highlightColor=(255,77,166,255);
    separatorColor=(255,255,255,255);
    enableWaveBlending=0;
};

```

2018100721021700-db1426d1dfd034027cecde9c2dd914b8
2018100721015700-db1426d1dfd034027cecde9c2dd914b8

**Comment by danielbernard** [comment]
Sure, I assume you want the text shifted left so the battery icon is aligned with the right edge of the time and "Theme Menu"?

**Comment by yellows8** [comment]
yes

**Comment by danielbernard** [comment]
Screenshot with latest:
2018100721502200-db1426d1dfd034027cecde9c2dd914b8

**Comment by nwert** [comment]
Is it just me or is the protrusion in the battery icon one pixel off?

**Comment by danielbernard** [comment]
You're absolutely right - it isn't originally, but when it's being scaled down it's off.
I'll figure it out tonight when I get home, and make sure that's fixed.

**Comment by danielbernard** [comment]
And actually, the charge icon looks jagged when scaled down that far too. Hmm. I'll make ones that scale better.

**Comment by yellows8** [comment]
Or don't scale it in the first place.

**Comment by danielbernard** [comment]
Could just make a tiny icon. I was intending to keep a larger size so that it could scale to the platform if desired, but just storing a smaller icon is probably more reliable

**Comment by danielbernard** [comment]
Here's with rescaled icons

2018100820002200-db1426d1dfd034027cecde9c2dd914b8
2018100820013200-db1426d1dfd034027cecde9c2dd914b8

File: common/common.h [hunk 2]
```diff
@@ -58,11 +58,13 @@ typedef union {
 #include "math.h"
 #include "theme.h"
 #include "message-box.h"
+#include "power.h"
 
 void menuStartupPath(void);
 void menuStartup(void);
 void themeMenuStartup(void);
 void menuLoop(void);
+void menuExit(void);
```
Comment (yellows8) [inline common/common.h]: Doesn't exist anymore.

File: nx_main/nx_power.c [hunk 3]
```diff
@@ -0,0 +1,37 @@
+#include <switch.h>
+#include "../common/common.h"
+
+static bool psmInitialized;
+
+void powerGetDetails(uint32_t *batteryCharge, bool *isCharging) {
```
Comment (yellows8) [inline nx_main/nx_power.c]: This should return a bool indicating whether the the output is valid (when psmInitialized block runs), caller shouldn't display info in that case.

File: nx_main/nx_power.c [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    ChargerType charger = ChargerType_None;
+
+    *isCharging = false;
+    *batteryCharge = 0;
+
+    if (psmInitialized)
+    {
+        psmGetBatteryChargePercentage(batteryCharge);
```
Comment (yellows8) [inline nx_main/nx_power.c]: Error checks missing.

File: nx_main/nx_power.c (hunk 3 above)
Comment (yellows8) [reply nx_main/nx_power.c]: return false

File: common/common.h (hunk 2 above)
Comment (danielbernard) [reply common/common.h]: Apologies, thought I removed that. Good catch.

**Comment by danielbernard** [comment]
Hold on further review, I goofed up the way I handled setting failure conditions for nx_power's powerGetDetails

**Comment by danielbernard** [comment]
Should be resolved now

**Pull request merged**